app/com/linkedin/drelephant/exceptions/core/ExceptionFingerprintingSpark.java

```diff
@@ -0,0 +1,368 @@
+      if (data[idx].contains(HOSTNAME_PATTERN)) {
+        amAddress = data[idx].split(HOSTNAME_PATTERN)[0] + HOSTNAME_PATTERN + nodeManagerPort;
+      } else if (data[idx].toLowerCase().contains(CONTAINER_PATTERN)) {
+        containerID = data[idx];
+      } else if (idx == data.length - 1) {
+        userName = data[idx];
+      }
+    }
+    return amAddress + URL_PATTERN + containerID + URL_PATTERN + containerID + URL_PATTERN + userName;
+  }
+
+  private void driverLogProcessingForException(BufferedReader in, List<ExceptionInfo> exceptions) throws IOException {
+    String inputLine;
+    while ((inputLine = in.readLine()) != null) {
+      if (inputLine.length() <= THRESHOLD_LOG_LINE_LENGTH.getValue() && isExceptionContains(inputLine)) {
+        if (debugEnabled) {
+          logger.debug(" ExceptionFingerprinting " + inputLine + "\t" + inputLine);
+        }
+        String exceptionName = inputLine;
+        int stackTraceLine = NUMBER_OF_STACKTRACE_LINE.getValue();
+        StringBuffer stackTrace = new StringBuffer();
+        while (stackTraceLine >= 0 && inputLine != null) {
+          stackTrace.append(inputLine);
+          stackTraceLine--;
+          inputLine = in.readLine();
+        }
+        addExceptions((exceptionName + "" + stackTrace).hashCode(), exceptionName, stackTrace.toString(),
+            ExceptionInfo.ExceptionSource.DRIVER, exceptions);
+        if (inputLine == null) {
+          break;
+        }
+      }
+    }
+  }
+
+  @Override
+  public LogClass classifyException(List<ExceptionInfo> exceptionInformation) {
+    if (exceptionInformation != null && exceptionInformation.size() > 0) {
+      Classifier classifier = ClassifierFactory.getClassifier(ClassifierType.RULE_BASE_CLASSIFIER);
+      classifier.preProcessingData(exceptionInformation);
+      LogClass logClass = classifier.classify(exceptionInformation);
+      return logClass;
+    }
+    return null;
+  }
+
+  @Override
+  public boolean saveData(String jobExecId) throws Exception {
+    JobExecution jobExecution = JobExecution.find.where().eq(JobExecution.TABLE.jobExecId, jobExecId).findUnique();
+    if (jobExecution == null) {
+      logger.error(" Job Execution with following id doesn't exist " + jobExecId);
+      throw new Exception("Job execution with " + jobExecId + " doesn't exist");
+    } else {
+      TuningJobDefinition tuningJobDefinition = TuningJobDefinition.find.where()
+          .eq(TuningJobDefinition.TABLE.job + "." + JobDefinition.TABLE.jobDefId, jobExecution.job.jobDefId)
+          .findUnique();
+      if (tuningJobDefinition == null) {
+        logger.error("Job definition with following job execution id doesnt' exisit. " + jobExecId);
+        throw new Exception("Job definition with following job execution id doesnt' exisit. " + jobExecId);
+      } else {
+        if (!tuningJobDefinition.autoApply) {
+          logger.info(" Auto tuning is not enabled on the job");
```

Comment:
Debug.

Reply:
let it be info  , as this help us to do the root cause analysis if something fails (as we are launching with info mode and not debug ,it will be difficult to do debug if this log is not there )

Reply:
OK. Would be good to add some context here like Job Definition ID.